bytecomp/lambda.ml

```diff
@@ -600,55 +600,80 @@ let rec subst s lam =
+                         List.map (subst_case update_env s) sw.sw_consts;
+                       sw_blocks =
+                         List.map (subst_case update_env s) sw.sw_blocks;
+                       sw_failaction =
+                         subst_opt update_env s sw.sw_failaction; },
               loc)
   | Lstringswitch (arg,cases,default,loc) ->
       Lstringswitch
-        (subst s arg,List.map (subst_strcase s) cases,subst_opt s default,loc)
-  | Lstaticraise (i,args) ->  Lstaticraise (i, subst_list s args)
+        (subst update_env s arg,
+         List.map (subst_strcase update_env s) cases,
+         subst_opt update_env s default,
+         loc)
+  | Lstaticraise (i,args) ->  Lstaticraise (i, subst_list update_env s args)
   | Lstaticcatch(body, (id, params), handler) ->
-      Lstaticcatch(subst s body, (id, params),
-                   subst (remove_list params s) handler)
+      Lstaticcatch(subst update_env s body, (id, params),
+                   subst update_env (remove_list params s) handler)
   | Ltrywith(body, exn, handler) ->
-      Ltrywith(subst s body, exn, subst (Ident.Map.remove exn s) handler)
-  | Lifthenelse(e1, e2, e3) -> Lifthenelse(subst s e1, subst s e2, subst s e3)
-  | Lsequence(e1, e2) -> Lsequence(subst s e1, subst s e2)
-  | Lwhile(e1, e2) -> Lwhile(subst s e1, subst s e2)
+      Ltrywith(subst update_env s body, exn,
+               subst update_env (Ident.Map.remove exn s) handler)
+  | Lifthenelse(e1, e2, e3) ->
+      Lifthenelse(subst update_env s e1,
+                  subst update_env s e2,
+                  subst update_env s e3)
+  | Lsequence(e1, e2) -> Lsequence(subst update_env s e1, subst update_env s e2)
+  | Lwhile(e1, e2) -> Lwhile(subst update_env s e1, subst update_env s e2)
   | Lfor(v, lo, hi, dir, body) ->
-      Lfor(v, subst s lo, subst s hi, dir,
-        subst (Ident.Map.remove v s) body)
+      Lfor(v, subst update_env s lo, subst update_env s hi, dir,
+        subst update_env (Ident.Map.remove v s) body)
   | Lassign(id, e) ->
       assert(not (Ident.Map.mem id s));
-      Lassign(id, subst s e)
+      Lassign(id, subst update_env s e)
   | Lsend (k, met, obj, args, loc) ->
-      Lsend (k, subst s met, subst s obj, subst_list s args, loc)
-  | Levent (lam, evt) -> Levent (subst s lam, evt)
-  | Lifused (v, e) -> Lifused (v, subst s e)
-and subst_list s l = List.map (subst s) l
-and subst_decl s (id, exp) = (id, subst s exp)
-and subst_case s (key, case) = (key, subst s case)
-and subst_strcase s (key, case) = (key, subst s case)
-and subst_opt s = function
+      Lsend (k,
+             subst update_env s met,
+             subst update_env s obj,
+             subst_list update_env s args,
+             loc)
+  | Levent (lam, evt) ->
+      let lev_env =
+        Ident.Map.fold (fun id _ env ->
+          match Env.find_value (Path.Pident id) evt.lev_env with
+          | exception Not_found -> env
+          | vd -> update_env id vd env
```

Comment:
I think that in this branch `update_env` should be called on a version of `env` where the `id` binding has been removed. Otherwise there is no way to remove identifiers from the environment when performing a substitution.